dGame/dPropertyBehaviors/ControlBehaviors.cpp

```diff
@@ -107,292 +92,68 @@ void ModelTypeChanged(AMFArrayValue* arguments, ModelComponent* ModelComponent)
 	//TODO Update the model type here
 }
 
-void ToggleExecutionUpdates() {
+void ControlBehaviors::ToggleExecutionUpdates() {
 	//TODO do something with this info
 }
 
-void AddStrip(AMFArrayValue* arguments) {
-	auto* strip = arguments->FindValue<AMFArrayValue>("strip");
-	if (!strip) return;
-
-	auto* actions = strip->FindValue<AMFArrayValue>("actions");
-	if (!actions) return;
-
-	auto* uiArray = arguments->FindValue<AMFArrayValue>("ui");
-	if (!uiArray) return;
-
-	auto* xPositionValue = uiArray->FindValue<AMFDoubleValue>("x");
-	if (!xPositionValue) return;
-
-	double xPosition = xPositionValue->GetDoubleValue();
-
-	auto* yPositionValue = uiArray->FindValue<AMFDoubleValue>("y");
-	if (!yPositionValue) return;
-
-	double yPosition = yPositionValue->GetDoubleValue();
-
-	STRIPID stripID = GetStripIDFromArgument(arguments);
-
-	BEHAVIORSTATE stateID = GetBehaviorStateFromArgument(arguments);
-
-	uint32_t behaviorID = GetBehaviorIDFromArgument(arguments);
-
-	std::string type = "";
-	std::string valueParameterName = "";
-	std::string valueParameterString = "";
-	double valueParameterDouble = 0.0;
-	for (uint32_t position = 0; position < actions->GetDenseValueSize(); position++) {
-		auto* actionAsArray = actions->GetValueAt<AMFArrayValue>(position);
-		if (!actionAsArray) continue;
-
-		for (auto& typeValueMap : actionAsArray->GetAssociativeMap()) {
-			if (typeValueMap.first == "Type") {
-				if (typeValueMap.second->GetValueType() != AMFValueType::AMFString) continue;
-
-				type = static_cast<AMFStringValue*>(typeValueMap.second)->GetStringValue();
-			} else {
-				valueParameterName = typeValueMap.first;
-				// Message is the only known string parameter
-				if (valueParameterName == "Message") {
-					if (typeValueMap.second->GetValueType() != AMFValueType::AMFString) continue;
-					valueParameterString = static_cast<AMFStringValue*>(typeValueMap.second)->GetStringValue();
-				} else {
-					if (typeValueMap.second->GetValueType() != AMFValueType::AMFDouble) continue;
-					valueParameterDouble = static_cast<AMFDoubleValue*>(typeValueMap.second)->GetDoubleValue();
-				}
-			}
-		}
-		// modelComponent->AddStrip(stateID, stripID, type, behaviorID, valueParameterName, valueParameterString, valueParameterDouble, "", xPosition, yPosition);
-		type.clear();
-		valueParameterName.clear();
-		valueParameterString.clear();
-		valueParameterDouble = 0.0;
-	}
-	// RequestUpdatedID(behaviorID);
+void ControlBehaviors::AddStrip(AMFArrayValue* arguments) {
+	AddStripMessage addStripMessage;
+	addStripMessage.Parse(arguments);
```

Comment:
If you call parse immediately on a new instance and don't use it anywhere else, just make it into a constructor or a static factory function

Reply:
(otherwise you may have some code down the line that calls parse on an already initialized instance and cause heisenbugs if there some old data not properly reset)

Reply:
Would the static factory function take in an AMFArrayValue* and then return a AddStripMessage in this case?

Reply:
yes, exactly